provider/gateway/router.go

```diff
@@ -116,9 +150,152 @@ func leaseServiceStatusHandler(log log.Logger, cclient cluster.ReadClient) http.
+		if err != nil {
+			if _, ok := err.(websocket.HandshakeError); !ok {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+			return
+		}
+
+		go wsLogWriter(ws, wsLogsConfig{
+			lid:       requestLeaseID(r),
+			service:   service,
+			follow:    requestLogFollow(r),
+			tailLines: requestLogTailLines(r),
+			log:       log,
+			client:    cclient,
+		})
+	}
+}
+
+func wsLogWriter(ws *websocket.Conn, cfg wsLogsConfig) {
+	pingTicker := time.NewTicker(pingPeriod)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer func() {
+		pingTicker.Stop()
+		cancel()
+		_ = ws.Close()
+	}()
+
+	logs, err := cfg.client.ServiceLogs(ctx, cfg.lid, cfg.service, cfg.follow, cfg.tailLines)
+	if err != nil {
+		cfg.log.Error("couldn't fetch logs: %s", err.Error())
+		err = ws.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocketInternalServerErrorCode, err.Error()))
+		if err != nil {
+			cfg.log.Error("couldn't push control message through websocket: %s", err.Error())
+		}
+		return
+	}
+
+	if len(logs) == 0 {
+		_ = ws.WriteMessage(
+			websocket.CloseMessage,
+			websocket.FormatCloseMessage(websocketInternalServerErrorCode, "service "+cfg.service+" does not have running pods"))
+		return
+	}
+
+	_ = ws.SetReadDeadline(time.Now().Add(pongWait))
+	ws.SetPongHandler(func(string) error {
+		_ = ws.SetReadDeadline(time.Now().Add(pongWait))
+		return nil
+	})
+
+	go func() {
+		for {
+			if _, _, e := ws.ReadMessage(); e != nil {
+				if _, ok := e.(*websocket.CloseError); !ok {
+					_ = ws.Close()
+				}
+				break
+			}
+		}
+	}()
+
+	var scanners sync.WaitGroup
+	var scannersDone sync.WaitGroup
+	var scannersStarted sync.WaitGroup
+
+	logCh := make(chan cluster.ServiceLogMessage)
+
+	scanners.Add(len(logs))
+	scannersStarted.Add(len(logs))
+
+	for _, lg := range logs {
+		go func(name string, scan *bufio.Scanner) {
+			defer scanners.Done()
+			scannersStarted.Done()
+
+			for scan.Scan() {
+				logCh <- cluster.ServiceLogMessage{
+					Name:    name,
+					Message: scan.Text(),
+				}
+			}
+		}(lg.Name, lg.Scanner)
+	}
+
+	scannersStarted.Wait()
+
+	scannersDone.Add(1)
+	go func() {
+		defer scannersDone.Done()
+		scanners.Wait()
+		_ = ws.Close()
+	}()
+
+	alive := true
+
+	for alive {
+		select {
+		case line := <-logCh:
+			stream, err := json.Marshal(line)
+			if err != nil {
+				alive = false
+				break
+			}
+
+			if err = ws.WriteMessage(websocket.TextMessage, stream); err != nil {
+				alive = false
+				break
+			}
+		case <-pingTicker.C:
+			if err = ws.WriteMessage(websocket.PingMessage, []byte{}); err != nil {
+				alive = false
+				break
+			}
+		}
+	}
+
+	for i := range logs {
+		_ = logs[i].Stream.Close()
 	}
+
+	scannersDone.Wait()
+
+	// drain logs channel in separate goroutine to unblock seeders waiting for write space
+	scannersStarted.Add(1)
+	go func() {
+		defer scannersStarted.Done()
+		for range logCh {
+		}
+	}()
+
+	close(logCh)
+	scannersStarted.Wait()
```

Comment:
Again let's forgo the ping stuff for now.

1. I don't think we need `scannersStarted`.
1. `logCh` doesn't need to be closed.
1. drain loop seems off; the writers are already in their own threads.
1. this should probably be in the same thread as the initial http handler.

```go
for _, lg := range logs {
  go func(name string, scan *bufio.Scanner) {
    defer scanners.Done()
    defer scan.Close()
    for scan.Scan() && ctx.Err() == nil {
       // ...
    }
  }()
}

donech := make(chan struct{})
go func() {
  scanners.Wait()
  close(donech)
}()
// ...

for alive {
  select {
    case line := <- logCh:
      ws.WriteJSON(line)
      // if err alive = false
    case <-donech:
      alive = false
  }
}

// tell all the log readers to stop.
cancel()

ws.WriteMessage(CloseMessage,...)

drain:
for {
  select {
    case <- donech:
      break drain
    case <- logCh:
  }
}
```

Reply:
you were right to close the `*bufio.Scanner`s in the main/outer thread - `cancel()` won't shut down the logs coming from kube, you have to `Close()` the bufio.

Reply:
oh wait that's why i added the `ctx.Err() == nil` condition.  never mind.